### PR TITLE
SOF-1743 Don't generate all keyboard bindings on every single update

### PR DESCRIPTION
**Pull request by LindvedKrvang**
'SystemKeybindings' are only updated when there is a Rundown change. 'ActionTriggerKeybindings' are only updated when there is an update to ActionTriggers or Actions.

Renamed `KeyBindingFactory` to `SystemKeyBindingFactory`.

Updated deafult value for `mappedKeys` when creating an ActionTrigger.

### CODE REVIEW

File: src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts [hunk 1]
```diff
@@ -94,26 +96,16 @@ export class ActionTriggerProducerKeyBindingService implements KeyBindingService
...
-      return keyBindings
-    }
-
+  private updateActionTriggerKeybindings(): void {
+    const keyBindings: KeyBinding[] = []
     this.actionTriggersWithAction.forEach((action: Tv2Action, actionTrigger: ActionTrigger<KeyboardTriggerData>) => {
       keyBindings.push(this.createBinding(action, actionTrigger, this.rundown!.id))
     })
-    return keyBindings
+    this.actionTriggerKeyBindings = keyBindings
```
Comment (KvelaGorrrrnio) [inline src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts]: The complexity of this can be reduced with by making this into a single expression either with `Object.entries`, `[...mapVariable]` or `Array.from`. I have provided a suggestion with `Array.from`.
```typescript
this.actionTriggerKeyBindings = Array.from(
  this.actionTriggersWithAction,
  ([actionTrigger, action]) => this.createBinding(action, actionTrigger, this.rundown!.id)
)
```

File: src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts [hunk 2]
```diff
@@ -185,7 +178,25 @@ export class ActionTriggerProducerKeyBindingService implements KeyBindingService
...
+
+    if (this.rundown.mode === RundownMode.INACTIVE) {
+      this.systemKeyBindings = this.systemKeyBindingFactory.createInactiveRundownKeyBindings(this.rundown)
+      return
+    }
+
+    this.systemKeyBindings =
+      this.rundown.mode === RundownMode.ACTIVE
+        ? this.systemKeyBindingFactory.createActiveRundownKeyBindings(this.rundown)
+        : this.systemKeyBindingFactory.createRehearsalRundownKeyBindings(this.rundown)
```
Comment (KvelaGorrrrnio) [inline src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts]: This seems like the perfect case for a switch statement :D
To reduce the complexity, the generation and assignment can be separated to only have one place where we assign. E.g.

```typescript
private updateSystemKeyBindings(): void {
  this.systemKeyBindings = this.createSystemKeyBindings()
}

private createSystemKeyBindings(): KeyBinding[] {
  if (!this.rundown) {
    return []
  }

  switch (this.rundown.mode) {
    case RundownMode.INACTIVE:
      return this.systemKeyBindingFactory.createInactiveRundownKeyBindings(this.rundown)
    case RundownMode.ACTIVE:
      return this.systemKeyBindingFactory.createActiveRundownKeyBindings(this.rundown)
    case RundownMode.REHEARSAL:
      return this.systemKeyBindingFactory.createRehearsalRundownKeyBindings(this.rundown)
  }
}

File: src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ export class EditKeyboardMappingComponent implements OnInit {
       label: [this.keyboardMapping?.actionTrigger.data.label ?? ''],
       keys: [this.keyboardMapping?.actionTrigger.data.keys, [Validators.required]],
       triggerOn: [KeyEventType.RELEASED, [Validators.required]],
-      mappedToKeys: [this.keyboardMapping?.actionTrigger.data.mappedToKeys],
+      mappedToKeys: [this.keyboardMapping?.actionTrigger.data.mappedToKeys ?? []],
```
Comment (KvelaGorrrrnio) [inline src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: Why is this added?

Comment (LindvedKrvang) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: Without it, the InputField for MappedKeys would display the "x" to clear the field, but the field was already "clear". So this makes the field behave correctly.

File: src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts (hunk 1 above)
Comment (LindvedKrvang) [reply src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts]: I wasn't aware of the override of `Array.from()`. I'm a fan :) 
It has been updated.

File: src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts (hunk 2 above)
Comment (LindvedKrvang) [reply src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts]: Updated

**Review by RasmusAlbrektsen — APPROVED** [review]

File: src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts (hunk 3 above)
Comment (KvelaGorrrrnio) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: I just tried removing this and nothing changed about when the clear button was shown in the keyboard input field. But it changes the runtime type of mappedToKeys from `Keys | undefined` to `Keys | string[] | undefined`.

Comment (LindvedKrvang) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: Sorry, I got it a bit messed up. If you don't have the `?? []` then when you create a new KeyboardMapping, the `mappedKeys` is going to have the value `null`, which is not allowed in our Zod parser.

Try creating a new KeyboardMapping and check your console.

Comment (KvelaGorrrrnio) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: If we use `[]` as the default value for the mappedToKeys input field, then we should check for an empty array and convert it to `undefined` when saving.
Another solution is to make the default value of the KeyboardInputComponent `undefined` instead of `null`.

Comment (LindvedKrvang) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: The default value of KeyboardInputComponent is already `undefined`. It's Angular forms that puts it as `null` when you don't provide a value.
If you give the input field a value and then clear it with the "x", then the value is going to be `undefined`.

Comment (KvelaGorrrrnio) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: The exposed default value should be `undefined` instead of `null`, such that the input field only provides valid values.

Comment (LindvedKrvang) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: As mentioned in my comment above. The default value in the InputField is already `undefined`. It's Angular forms that changes the value to `null` when you don't provide a value to the `FormControl`.

Comment (KvelaGorrrrnio) [reply src/app/settings/keyboard-mapping-settings/components/edit-keyboard-mapping/edit-keyboard-mapping.component.ts:36]: We'll leave it as is and I'll create a task for fixing it.

**Review by KvelaGorrrrnio — APPROVED** [review]

**Review by k-primeholding-dim — APPROVED** [review]

**Review by vivanov90 — APPROVED** [review]
